### DISCUSSION TITLE
NPC Poison/Venom Tracker

### DISCUSSION THREAD

**Discussion by Subzotic** [runelite/runelite#16269] [Ideas]
With the new Poison/Venom changes, attackable NPCs will now continue to take Poison/Venom damage even if they de-aggro.
Specific account builds utilize poison/venom to minimize XP gained in combat skills.

I suggest a plugin that highlights (maybe have overlay options) NPCs the player has inflicted with poison or venom. That way they are easier to keep track of and the player won't have to keep trying to poison/venom something that has already been inflicted.

This would **ONLY** be for attackable NPCs that can be poisoned/venomed, **NOT** players.

I would love to work with someone on this as I have limited coding knowledge. I was utilizing ChatGPT on reference code for something like this and it looked fairly promising. 

_Other discussions that brought up something similar have not really led to anything, which is why I wanted to submit this. Thanks!_

**Comment by Subzotic** [comment]
Here is my repository with some of the code I'm working on: https://github.com/Subzotic/NPCPoisonVenomTracker.git

So far it is showing up in the Plugin Hub but it doesn't seem to do anything. Any assistance would be appreciated!

**Comment by salverrs** [comment]
Hey, I'd love to help out with this! I've already made some fixes and additions to the code that I think will be useful, I will create a PR on your repo tomorrow and provide some more info on what I've changed, It's still only a basic implementation though.

**Reply by Subzotic** [reply]
Thanks! I just saw the PR and tested it out. It's working great and I have not run into any real issues so far. I'm still really new to coding and you just came in to the rescue! I have a lot more to learn and looking over your code has connected quite a few dots I was missing. 

I **REALLY** appreciate the response, thank you so much!

In terms of input, I think having the overhead text be an optional setting in the config window would be better. Also, I noticed that the highlight disappears once an NPC is out of render range. Would it be possible to have the highlight "remember" the specific NPC once back in render range? Then having other overlay options could be useful as well for personalization. (I'll try my hand at adding these now that you put in a good foundation.)

I feel bad that you had to completely rewrite my code for my own repo so I am really thankful that you gave your time to this. I hope one day I can do what you did and fulfill random Github RuneLite Plugin requests for those in need. haha

So how does this work now? I will be merging the PR but I want to make sure you are credited appropriately. Still fairly new to this, sorry.

**Reply by salverrs** [reply]
No problem! I kinda forgot to consider NPCs moving out of rendering distance. I think NPC references change when they leave and return into render distance so that's why it currently breaks when doing so, the plugin doesn't realise it's the same NPC anymore. In order to fix this, I think you'll need to change any 'NPC' object references to 'NPCComposition' instead, which can be accessed through an NPC instance. This is because NPCComposition references don't change even when they aren't being rendered anymore. Also, I realised the event 'NpcDespawned' is also fired when the NPC stops rendering, so the method 'onNpcDespawned' should be removed because it's technically incorrect to stop tracking NPCs in those scenarios.

If you have any trouble implementing this, feel free to message me with any questions. I'm also happy to create another PR with the mentioned changes if that's something you'd prefer. Don't worry about crediting me, just happy to help :)

**Reply by Subzotic** [reply]
Gotcha, I'll see what I can do. Do you have any references or recommendations for learning more about coding RuneLite plugins? I think the main issue I have is knowing what the specific dependencies are that I can call on. But an overall overview would be helpful as well.

I will let you know if I run into any issues, thanks again!

**Reply by Chadillac406** [reply]
great project! i was testing it and maybe theres a way to add the overlay upon initial hit? good for uses with trident/BP with serp!
